src/jit/inlinepolicy.cpp

```diff
@@ -1113,11 +1396,93 @@ void DiscretionaryPolicy::DetermineProfitability(CORINFO_METHOD_INFO* methodInfo
     }
 
     // Make some additional observations
-    m_ArgCount = methodInfo->args.numArgs;
-    m_LocalCount = methodInfo->locals.numArgs;
-    m_ReturnType = methodInfo->args.retType;
 
-    // delegate to LegacyPolicy for now
+    CORINFO_SIG_INFO& locals = methodInfo->locals;
+    m_LocalCount = locals.numArgs;
+
+    CORINFO_SIG_INFO& args = methodInfo->args;
+    const unsigned argCount = args.numArgs;
+    m_ArgCount = argCount;
+
+    const unsigned pointerSize = sizeof(void*);
+    unsigned i = 0;
+
+    // Implicit arguments
+
+    const bool hasThis = args.hasThis();
+
+    if (hasThis)
+    {
+        m_ArgType[i] = CORINFO_TYPE_CLASS;
+        m_ArgSize[i] = pointerSize;
+        i++;
+        m_ArgCount++;
+    }
+
+    const bool hasTypeArg = args.hasTypeArg();
+
+    if (hasTypeArg)
+    {
+        m_ArgType[i] = CORINFO_TYPE_NATIVEINT;
+        m_ArgSize[i] = pointerSize;
+        i++;
+        m_ArgCount++;
+    }
+
+    // Explicit arguments
+
+    unsigned j = 0;
+    CORINFO_ARG_LIST_HANDLE argListHandle = args.args;
+    COMP_HANDLE comp = m_RootCompiler->info.compCompHnd;
+
+    while ((i < MAX_ARGS) && (j < argCount))
+    {
```

Comment:
Believe this is correct as is -- if an inlinee has 4 args, we'll track them all, in array slots 0, 1, 2, 3. So the max arg index is one less than MAX_ARGS.